eta/core/serial.py

```diff
@@ -406,24 +407,424 @@ def _recurse(v, reflective):
+
+        Args:
+            element: an instance of `_ELE_CLS`
+
+        Returns:
+            the key for the element
+        '''
+        return getattr(element, self._ELE_KEY_ATTR)
+
+    @property
+    def size(self):
+        '''Returns the number of elements in the set.'''
+        return len(self)
+
+    def clear(self):
+        '''Deletes all elements from the set.'''
+        setattr(self, self._ELE_ATTR, OrderedDict())
+
+    def copy(self):
+        '''Returns a deep copy of the set.
+
+        Returns:
+            a Set
+        '''
+        return copy.deepcopy(self)
+
+    def get_keys(self):
+        '''Returns the set of keys for the elements of the set.'''
+        return set(self.__elements__)
+
+    def add(self, element):
+        '''Adds an element to the container.
+
+        Args:
+            element: an instance of `_ELE_CLS`
+        '''
+        key = self.get_key(element) or str(uuid4())
+        self.__elements__[key] = element
+
+    def add_set(self, set_):
+        '''Adds the elements in the given set to this set.
+
+        Args:
+            set_: a Set
+        '''
+        self.__elements__.update(set_.__elements__)
+
+    def add_iterable(self, elements):
+        '''Adds the elements in the given iterable to the set.
+
+        Args:
+            elements: an iterable of `_ELE_CLS` objects
+        '''
+        for element in elements:
+            self.add(element)
+
+    def filter_elements(self, filters, match=any):
+        '''Removes elements that don't match the given filters from the set.
+
+        Args:
+            filters: a list of functions that accept elements and return
+                True/False
+            match: a function (usually `any` or `all`) that accepts an iterable
+                and returns True/False. Used to aggregate the outputs of each
+                filter to decide whether a match has occurred. The default is
+                `any`
+        '''
+        elements = self._filter_elements(filters, match)
+        setattr(self, self._ELE_ATTR, elements)
+
+    def delete_keys(self, keys):
+        '''Deletes the elements from the set with the given keys.
+
+        Args:
+            keys: an iterable of keys of the elements to delete
+        '''
+        for key in keys:
+            del self.__elements__[key]
+
+    def keep_keys(self, keys):
+        '''Keeps only the elements in the set with the given keys.
+
+        Args:
+            keys: an iterable of keys of the elements to keep
+        '''
+        elements = self._get_elements(keys)
+        setattr(self, self._ELE_ATTR, elements)
+
+    def extract_keys(self, keys):
+        '''Creates a new set having only the elements with the given keys.
+
+        Args:
+            keys: an iterable of keys of the elements to keep
+
+        Returns:
+            a Set
+        '''
+        elements = copy.deepcopy(self._get_elements(keys))
```

Comment:
So the assumption for `Container` and `Set` is that subclasses do not add there own attributes, i.e. `Container`s and `Set`s are containers and sets and nothing more?

Or is it just `extract_keys`, etc. do not copy those attributes over to returned instance...

Reply:
Good point. I was trying to avoid copying the entire object, but that's a bit safer in case subclasses have other stuff in them. Fixed